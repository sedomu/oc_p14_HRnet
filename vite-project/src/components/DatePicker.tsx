import * as React from "react";
import { ChevronDownIcon } from "lucide-react";

import { Button } from "@/components/ui/button";
import { Calendar } from "@/components/ui/calendar";

import {
    Popover,
    PopoverContent,
    PopoverTrigger,
} from "@/components/ui/popover";

type DatePickerProps = {
    name: string;
    dataTestId: string;
};

export function DatePicker({ name, dataTestId }: DatePickerProps) {
    const [open, setOpen] = React.useState(false);
    const [date, setDate] = React.useState<Date | undefined>(undefined);

    const toISODateString = (date: Date) => {
        const year = String(date.getFullYear());
        const month = String(date.getMonth() + 1).padStart(2, "0");
        const day = String(date.getDate()).padStart(2, "0");

        return `${year}-${month}-${day}`;
    };

    return (
        <>
            <Popover open={open} onOpenChange={setOpen}>
                <PopoverTrigger asChild>
                    <Button
                        variant="outline"
                        id="date"
                        className="w-full justify-between font-normal"
                    >
                        <span>{date ? date.toLocaleDateString() : ""}</span>
                        <ChevronDownIcon />
                    </Button>
                </PopoverTrigger>
                <PopoverContent
                    className="w-auto overflow-hidden p-0"
                    align="end"
                >
                    <Calendar
                        mode="single"
                        selected={date}
                        captionLayout="dropdown"
                        onSelect={(date) => {
                            setDate(date);
                            setOpen(false);
                        }}
                    />
                </PopoverContent>
            </Popover>
            <input
                type="hidden"
                name={name}
                value={date ? toISODateString(date) : ""}
                data-testid={dataTestId}
            />
        </>
    );
}
