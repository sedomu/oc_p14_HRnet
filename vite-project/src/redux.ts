import {configureStore, createSlice, type PayloadAction} from '@reduxjs/toolkit'
import {MOCKUP_DATA} from "@/config.ts";
import {employeesMockupData} from "@/data/employees_mockup.ts";

export type Employee = {
    firstName: string
    lastName: string
    dateOfBirth: string
    startDate: string
    street: string
    city: string
    state: string
    zipCode: number
    department: string
}

const initialState: Employee[] = []

if (MOCKUP_DATA){
    initialState.push(...employeesMockupData);
}

const employeesSlice = createSlice({
    name: 'employees',
    initialState: initialState,
    reducers: {
        addEmployee: (state, action: PayloadAction<Employee>) => {
            state.push(action.payload)
        }
    }
})

export const store = configureStore({
    reducer: {
        employees: employeesSlice.reducer
    }
})

export const { addEmployee } = employeesSlice.actions

export type RootState = ReturnType<typeof store.getState>;