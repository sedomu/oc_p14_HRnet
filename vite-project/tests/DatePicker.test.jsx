import { describe, it, expect } from "vitest";
import { DatePicker } from "@/components/DatePicker.js";
import { Provider } from "react-redux";
import { store } from "@/redux";
import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { useState } from "react";

describe("DatePicker", () => {
    it("should render correctly", () => {
        const { container } = render(
            <Provider store={store}>
                <DatePicker name={"test"} ariaLabel={""} dataTestId={"test"} />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });
    it("should display date in correct format", async () => {
        const TestComponent = () => {
            const [date, setDate] = useState(undefined);
            return (
                <DatePicker
                    name={"test"}
                    dataTestId={"test"}
                    ariaLabel={""}
                    value={date}
                    onChange={setDate}
                />
            );
        };

        render(
            <Provider store={store}>
                {/*<DatePicker name={"test"} dataTestId={"test"} />*/}
                <TestComponent />
            </Provider>
        );

        await userEvent.click(screen.getByRole("button"));
        const yearSelect = await document.querySelector(".rdp-years_dropdown");
        await userEvent.selectOptions(yearSelect, "2022");
        const monthSelect = await document.querySelector(
            ".rdp-months_dropdown"
        );
        await userEvent.selectOptions(monthSelect, "May");
        await userEvent.click(await screen.getByText("15"));

        expect(screen.getByTestId("test").value).toBe("2022-05-15");
    });
});
