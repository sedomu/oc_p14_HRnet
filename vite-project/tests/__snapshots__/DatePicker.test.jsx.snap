// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatePicker > should render correctly 1`] = `
<div>
  <button
    aria-controls="radix-«r0»"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full justify-between font-normal"
    data-slot="popover-trigger"
    data-state="closed"
    id="date"
    role="button"
    type="button"
  >
    <span />
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </button>
  <input
    data-testid="test"
    name="test"
    type="hidden"
    value=""
  />
</div>
`;
